stacker/hooks/aws_lambda.py

```diff
@@ -0,0 +1,318 @@
+import os
+import os.path
+import stat
+import logging
+import hashlib
+from StringIO import StringIO
+from zipfile import ZipFile, ZIP_DEFLATED
+
+import boto3
+import botocore
+import formic
+from troposphere.awslambda import Code
+
+from stacker.util import get_config_directory
+
+
+# UNIX file attributes are stored in the upper 16 bits in the external
+# attributes field of a ZIP entry
+ZIP_PERMS_MASK = (stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO) << 16
+
+logger = logging.getLogger(__name__)
+
+
+def _zip_files(files, root):
+    """
+    Generate a ZIP file in-memory containing the chosen files, specified
+    relative to the root (and stored with those same relative paths in the
+    ZIP).
+    Normalizes UNIX permissions to avoid any problems with Lambda execution.
+
+    Returns the contents as a string.
+    """
+    zip_data = StringIO()
+    with ZipFile(zip_data, 'w', ZIP_DEFLATED) as zip_file:
+        for fname in files:
+            zip_file.write(os.path.join(root, fname), fname)
+
+        # Fix file permissions to avoid any issues - only care whether a file
+        # is executable or not, choosing between modes 755 and 644 accordingly.
+        for zip_entry in zip_file.filelist:
+            perms = (zip_entry.external_attr & ZIP_PERMS_MASK) >> 16
+            if perms & stat.S_IXUSR != 0:
+                new_perms = 0755
+            else:
+                new_perms = 0644
+
+            if new_perms != perms:
+                logger.debug("lambda: fixing perms: %s: %o => %o",
+                             zip_entry.filename, perms, new_perms)
+                new_attr = ((zip_entry.external_attr & ~ZIP_PERMS_MASK) |
+                            (new_perms << 16))
+                zip_entry.external_attr = new_attr
+
+    contents = zip_data.getvalue()
+    zip_data.close()
+
+    return contents
+
+
+def _find_files(root, includes, excludes):
+    """
+    Generate a list of files relative to a root path, applying inclusion and
+    exclusion patterns. The documentation for the patterns can be found at:
+    http://www.aviser.asia/formic/doc/index.html
+    """
+
+    root = os.path.abspath(root)
```

Comment:
Done. I didn't end up using `full_path` because I don't want to call `get_config_directory` unless the path is relative, as that would force me to mock it in all the tests. I just call `expanduser` right when I retrieve the config key.